### ISSUE TITLE
given a source kubeconfig which has the 'insecure-skip-tls-verify: true' set, 'pinniped get kubeconfig' should also use this

### ISSUE THREAD

**Issue by lknite** [vmware-tanzu/pinniped#1252]
**What happened?**

I have setup external-dns to register my cluster kubeapi ips as fqdn.

To access them I have to set the insecure flag to true in the kubeconfig:
```
apiVersion: v1
clusters:
- cluster:
    #certificate-authority-data: ...
    server: https://clustername.k.home.net:6443
    insecure-skip-tls-verify: true
```

After generating a pinniped kubeconfig and trying to access a cluster it fails with an unknown certificate error.  This seems to be showing the 'insecure-skip-tls-verify' setting was lost.

**What did you expect to happen?**

I expected the kubeconfig generated by 'pinniped get kubeconfig' to also respect the 'insecure-skip-tls-verify' value and carry that through to its kubeconfig as needed.

**What is the simplest way to reproduce this behavior?**

1. set a dns entry to the kubeapi ip of a cluster
2. modify the kubeconfig you will be passing to 'pinniped get kubeconfig' setting the server url to the fqdn, instead of the usual ip address, also add the 'insecure-skip-tls-verify: true' option
3. run a 'kubectl get nodes' and see that it works
4. generate a pinniped kubeconfig using 'pinniped get kubeconfig' and passing it the modified kubeconfig
5. now set KUBECONFIG environment variable to point at the pinniped generated kubeconfig, after logging in note all connections fail with a certificate error

**In what environment did you see this bug?**
- Pinniped server version: ```not used, only using concierge```
- Pinniped client version:  ```version.Info{Major:"0", Minor:"17", GitVersion:"v0.17.0", GitCommit:"b458cd43b96813c848d42ad196dfc8b576c0488a", GitTreeState:"clean", BuildDate:"2022-05-06T19:17:03Z", GoVersion:"go1.18.1", Compiler:"gc", Platform:"linux/amd64"}```
- Pinniped container image (if using a public container image): ```n/a```
- Pinniped configuration (what IDP(s) are you using? what downstream credential minting mechanisms are you using?): ```keycloak```
- Kubernetes version (use `kubectl version`): ```Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.0+vmware.wcp.2", GitCommit:"d5bb17833505d15ce5f40815bb14fede978fe8c1", GitTreeState:"clean", BuildDate:"2021-08-14T16:46:51Z", GoVersion:"go1.16.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.12+vmware.1", GitCommit:"768a3bbd17406f20f51df36cbd12695c8293392d", GitTreeState:"clean", BuildDate:"2021-11-01T22:45:04Z", GoVersion:"go1.15.15", Compiler:"gc", Platform:"linux/amd64"}```
- Kubernetes installer & version (e.g., `kubeadm version`): ```kubeadm version: &version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.3", GitCommit:"aef86a93758dc3cb2c658dd9657ab4ad4afc21cb", GitTreeState:"clean", BuildDate:"2022-07-13T14:29:09Z", GoVersion:"go1.18.3", Compiler:"gc", Platform:"linux/amd64"}```
- Cloud provider or hardware configuration: ```n/a```
- OS (e.g: `cat /etc/os-release`): .
NAME="CentOS Stream"
VERSION="8"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="8"
PLATFORM_ID="platform:el8"
PRETTY_NAME="CentOS Stream 8"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:8"
HOME_URL="https://centos.org/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux 8"
REDHAT_SUPPORT_PRODUCT_VERSION="CentOS Stream"
- Kernel (e.g. `uname -a`): ```Linux itsku-root-w1 4.18.0-408.el8.x86_64 #1 SMP Mon Jul 18 17:42:52 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux```

**Comment by lknite** [comment]
Note, I was able to go back to my source kubeconfig, remove the insecure flag, uncomment the certificate-authority-data, and change the fqdn back to an ipaddress and pinniped is now working as expected.

By performing this workaround, it seems to have verified that pinniped is ignoring the 'insecure-skip-tls-verify' attribute rather than propagating it through.

**Comment by lknite** [comment]
Just updated pinniped to the latest.  Still needing to use the workaround for this one, and it is a bit of a hack, would be nice if pinniped could recognize what's happening and set things up as expected.  Adding post to hopefully keep this one alive.

**Comment by cfryanr** [comment]
Hi @lknite,

Can you please explain why you would like to use the insecure flag instead of using the certificate authority data? That seems insecure, since there would be no protection from person-in-the-middle attacks. You could be handing your credentials to anyone without protection.

Thanks,
Ryan

**Comment by lknite** [comment]
I'm using a tool to deploy the clusters, tanzu in this case, it deploys the clusters with an ip address.  What I want to do is access the cluster using an fqdn.  In general, the reason for this is so I can use gitops to deploy to a cluster.  Later, if the cluster is deleted and recreated the ip will probably be different, but as long as the gitops deployments are targeting an fqdn they will still target the right cluster and deploy once again.

Maybe you are thinking is that what I need to do is regenerate the certificates with the fqdn as a valid host so that the certificates will work.  That would be nice, though I'm not sure that's a very practical solution with tanzu at the moment.

I believe it's still going to use certificates, it's not just sending everything out over the line in plaintext, it's just not verifying that the target ip is matching the certificates ... instead i'm putting in my own value there, an fqdn, which i've setup in dns.

**Comment by lknite** [comment]
I suppose mitm is always a concern, in this case it is all onprem, but still you are right regarding that as a concern.

**Comment by lknite** [comment]
Maybe this is reason to not use tanzu, but we are still in the early stages...

'vsphere with tanzu' https://kb.vmware.com/s/article/86251

![image](https://user-images.githubusercontent.com/92187614/210662902-3f80b6bf-3cdc-4990-b25b-6c2d7bd021a2.png)

**Comment by cfryanr** [comment]
> I'm using a tool to deploy the clusters, tanzu in this case, it deploys the clusters with an ip address. What I want to do is access the cluster using an fqdn.

I see. This is getting into Tanzu's commercial integration of Pinniped, and starting to stray away from the features of the open source Pinniped project, but maybe this will be helpful to you... you can configure your TKG cluster's Pinniped to use an fqdn for the Pinniped Supervisor. I'm not sure if this is documented on the VMware web site or not, but I happen to know that it is possible. Maybe check with your VMware support team for details. It sounds like that's what you need, instead of a way to disable TLS verification in open source Pinniped?

**Comment by cfryanr** [comment]
Again, this is not an open source Pinniped concern, but again just trying to be helpful... here is a link to the doc about configuring an fqdn for Pinniped in TKG 1.6: see `pinniped.supervisor_svc_external_dns` in https://docs.vmware.com/en/VMware-Tanzu-Kubernetes-Grid/1.6/vmware-tanzu-kubernetes-grid-16/GUID-packages-update-addons.html#updating-package-configuration-2

**Comment by lknite** [comment]
Thank you.  We are using the tanzu integrated with vsphere, not the tkg.  In any case, I'm also not using the integrated pinniped, I'm using the open source version of pinniped.  I did ask the vmware support team for the ability to use fqdns with their clusters months ago and was told it wasn't possible, but they were ok with me using external-dns, so I set that up.

I can see your concern about enabling a feature which might be useful in non-prod, but also creates the possibility of someone mistakenly using it in prod, and if things lined up would promote the possibility of a mitm attack.  It's also safer to reference a cluster using an ip address than fqdn, which is a gray area, that some may chose to consider dns safe.  Folks using gitops will want to use the fqdn.

You've almost successfully talked me out of using Tanzu given that we want to use the fqdn but it's not possible to redo the certs to accept the fqdn rather than an ip address.  I'll put in a ticket asking them if there's any way to make this work.  I've heard rumor that their two separate versions of tanzu are going to go away and they'll just have the one code-base to support, at that point maybe this will be doable with the link you provided above.

In my home environment I've since figured out how to redo the certificates so am able to do this without using this --insecure flag now.  That was due to your resistance, so that worked out.

In my dev at work environment I have scripts to do this technique anyway manually, guess I'll keep doing that in the meantime, it's just inconvenient.  As a last attempt, cause it'd still be nice to have, kubernetes itself allows the flag ... I'm guessing you'd make the argument that's possibly irresponsible of them and they should remove it?  Your concerns are all warranted, and I don't disagree with you at all, and I'm fully in support of all things security related, but I do see the value of mirroring the functionality provided by kubernetes.

Maybe you could go ahead and implement the --insecure, then people could find it when searching the documents, and in the documents you could mention that you have to add '--insecure-override' to make --insecure work.  That way '--insecure' wouldn't work by default, but the customer could then google it when trying to get it work and find a web page that says something like, "We think it's a really bad idea for this flag to be used, it creates the possibility of a mitm attack, do not use this flag in production, add --insecure-override, to use this flag anyways against our recommendation".

**Comment by cfryanr** [comment]
Yes, the concern with having an insecure flag is that someone will use it in a prod environment which will open them up to all sorts of attacks. They might even use it on purpose in a prod environment, not fully understanding the implications, which we've seen before with other products. We tried to avoid all cases of using insecure flags in Pinniped for this reason. Although I can see your argument for supporting it too.

Of course, nothing stops you from overriding this behavior by editing the kubeconfig file, which you mentioned that you have already done.

> at that point maybe this will be doable with the link you provided above

I may have conflated two separate fqdn concerns when I shared that doc link. That's the doc for how to set the fqdn for the Pinniped Supervisor which is bundled into TKG. Reviewing this issue again, I think you were actually talking about the fqdn for the Kubernetes API server itself. That would be a different setting, although I think that is also allowed in TKG if I'm not mistaken.

I'm less familiar with tanzu integrated with vsphere, but I would hope that they also provide some way to configure the fqdn of the Kubernetes API server, or at least allow you to influence how its TLS cert gets created (by influencing the subject alt name, or similar). Or do they maybe have an option to influence which IP gets used, so that if the cluster is recreated it comes back at the same IP address?

If it were me personally, and I had to choose between security of avoiding person-in-the-middle attacks and the convenience of using an fqdn for my Kube API server, I would always choose security. But that's just me. :) And of course having the option to do both might be even better.

**Comment by lknite** [comment]
Well, I still like my proposed solution.  But, in the end it's your code to do as you wish, and I can't find any flaw in your wanting to protect people from themselves, hearing that just makes me trust using pinniped more.  I can learn to accept this decision.

I'll improve the work around script I've got and use that in dev, hoping that by the time we go full prod tanzu will be able to provide the fqdns we want, otherwise will have to abandon the strategy then.  fingers crossed.

Thank you very much.  I really appreciate that you took the time to speak on this subject.

**Comment by lknite** [comment]
After I learned more, I put together what I consider a much better solution to what I was proposing (and doesn't require ``--insecure`` to work).

It is possible to target an application to a cluster name using gitops (argocd), rather than URL, which achieves my goal above.

It's not possible with argocd by default, but I've created a controller which makes it possible:
https://github.com/lknite/daytwo

The 'argocd-register-controller' adds a label with the cluster name, this can then be used to target the cluster.  This means a cluster can be deleted and then created again with the same name and the application will reinstall again to the cluster.

argocd can target the clustername like so:
```
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: awesome-application
  namespace: argocd
spec:
  generators:
  - clusters:
      selector:
        matchLabels:
          argocd.argoproj.io/secret-type: cluster
    selector:
      matchLabels:
        metadata.annotations.daytwo.aarr.xyz/workload-cluster: "targetCluster"
```

Also, I created 'argocd-pinniped-controller' which I thought you might find entertaining.  It generates pinniped kubeconfigs automatically by looking at the argocd secrets & hosts a url through which all pinniped kubeconfigs can be accessed.  In production I deploy out a cluster-specific website which has a url back to the central pinniped kubeconfig hosting server.  So, no need to have a central git repo.

I'd recommend anyone ending up on this issue with a similar use case to implement something similar and avoid any hopes of using the ``--insecure`` flag even temporarily.
